### ISSUE TITLE
Look and feel defaults to Nimbus on Windows

### ISSUE THREAD

**Issue by tresf** [dorkbox/SystemTray#45]
With the latest Beta, Windows defaults to Nimbus Look and Feel (`CrossPlatformLookAndFeel`) whereas I would like to specifically set it to the Windows theme (`SystemLookAndFeel`)or something else of my choosing.

![image](https://cloud.githubusercontent.com/assets/6345473/22680582/ca9eaa68-ecd6-11e6-8de1-2a8cbd9ba30d.png)

**Comment by dorkbox** [comment]
Very strange -- I don't think I'm doing anything with the look and feel... I'll see what might be going on, but maybe it's from some testing code from @Vzor-?? I do know we were testing an issue with setting the L&F at one point.

**Comment by tresf** [comment]
> maybe it's from some testing code from @Vzor-

I'm using RC9 from your downloads area and @Vzor- has [zero commits on your branch](https://github.com/dorkbox/SystemTray/commits/master?author=Vzor-), so none I'm aware of.

**Comment by tresf** [comment]
Is there a chance you're [accidentally initializing a Swing component](https://coderanch.com/t/529311/java/Weird-System-Feel-behaviour) before `init()` is called (such as in a static method?)

**Comment by Vzor-** [comment]
Ill whip up a test case tonight to verify it is not in our code.

**Comment by Vzor-** [comment]
I pushed the test, this issue seems to be in dorkbox. https://github.com/Vzor-/SystemTray/commit/37712f5def347642752fc690d5c0d3705ba678d1 I'll help more after some sleep 😴

**Comment by Vzor-** [comment]
The issue occurs here https://github.com/Vzor-/SystemTray/blob/master/src/dorkbox/systemTray/SystemTray.java#L783

It resolves as a dorkbox.systemTray.swingUI._SwingTray but that is about the extent of what I can see with the debugger.

**Comment by Vzor-** [comment]
Dug a bit deeper, the issue is with the systemutil. The tray initializes and waits on a [setEnabled in MenuItem](https://github.com/Vzor-/SystemTray/blob/master/src/dorkbox/systemTray/MenuItem.java#L132). It then sits on [this invokeLater in _SystemTray](https://github.com/Vzor-/SystemTray/blob/master/src/dorkbox/systemTray/swingUI/_SwingTray.java#L69), at this point it is still a windows look and feel, but as soon as it enters the systemUtil, it runs some custom L&F logic and sets the look and feel to CUSTOM_LOOK_AND_FEEL, which is nimbus, even if I try to set it in test tray.

**Comment by Vzor-** [comment]
The menu looks strange under any l&f that isn't Nimbus 
![capture](https://cloud.githubusercontent.com/assets/10621658/22737893/2cde8958-edd3-11e6-888c-efe7c193297f.PNG)
may be a tough one to fix

**Comment by dorkbox** [comment]
I fixed the issue with the SystemUtil, so it's not set to a default value - I'll see what can be done about making the menu look nice

**Comment by tresf** [comment]
@dorkbox can you reference the commit so we can learn from the change and also close the bug report?

**Comment by dorkbox** [comment]
Not exactly. It's an internal library change (from "Nimbus" -> null) that isn't on github yet. I'm in the process of adding it to github, but its taking a while. As soon as it's added, I'll reference the commit so you can see it.

**Comment by dorkbox** [comment]
@tresf the commit that fixed this is : https://github.com/dorkbox/Utilities/commit/d8f0c4d8180625bdf0024c2df469a4eb752dbb00  I am still trying to figure out how to make the tray look nice WITHOUT the nimbus L&F (and also, with it...). Once that is resolved, I'll close this issue.

**Comment by tresf** [comment]
> I am still trying to figure out how to make the tray look nice WITHOUT the nimbus L&F (and also, with it...). Once that is resolved, I'll close this issue.

I don't understand this statement.  Swing looks fine and it should be on the developer to change L&F.  What's wrong with Swing?

**Comment by dorkbox** [comment]
I'm customizing the menu and menu items **slightly** (just the margins) which looked nice on Nimbus (even though I didn't realize it was Nimbus), and it looked wonky without it. IE: https://github.com/dorkbox/SystemTray/blob/master/src/dorkbox/systemTray/swingUI/AdjustedJMenuItem.java

**Comment by tresf** [comment]
> Which looked nice on Nimbus (even though I didn't realize it was Nimbus),

Couldn't this be the problem?  Swing defaults should be very close to the OS (if not, it's an Oracle bug that we should report or workaround), I would discourage changing anything that's not necessary.

**Comment by dorkbox** [comment]
Definitely, and I agree.

What I'm going to do is expose the ability for devs to customize their own menu/item components for a swing system tray. This would also provide the ability for very custom menu's on windows, like on windows 10.

**Comment by tresf** [comment]
To add more context to this statement... we've been using swing-drawn tray menus (JDK7, JDK8) for about two years without any issues.

![image](https://cloud.githubusercontent.com/assets/6345473/22831820/bc43296c-ef79-11e6-92e9-87e8e6eb200b.png)

> What I'm going to do is expose the ability for devs to customize their own menu/item components for a swing system tray.

I assume your end-goal is something like this, right?

![image](https://cloud.githubusercontent.com/assets/6345473/22831869/f0ecb426-ef79-11e6-9499-09e4ab17ee34.png)

It's a noble effort however keep in mind that of the icons available on my computer at the time of writing this, these are my observations:

* Bleeding-edge Office 365, still uses old Windows-style menus for the Outlook indicator and new mail indicator
* Volume indicator uses new Windows 10 style for left click and old Windows-style for right click
* Battery indicator uses new Windows 10 style for left click and old Windows-style for right click
* Network indicator uses new Windows 10 style for left click and old Windows-style for right click
* Inskspace and keyboard indicators use new Windows 10 style
* Java indicator uses old Windows-style
* Bluetooth uses old Windows-style
* McAfee uses old Windows-style
* USB/Safely remove uses old Window-style menus
* Pidgin uses old Windows-style menus

So it's about 50% using new Windows-style menus and the adoption rate is only Microsoft themselves.  No 3rd parties (at least on my machine) are making this switch.

**Comment by dorkbox** [comment]
> To add more context to this statement... we've been using swing-drawn tray menus (JDK7, JDK8) for about two years without any issues.

Good to know.

> I assume your end-goal is something like this, right?

Something like that, with this release providing the ability for simple changes (margins, font, color, etc), because for my projects that use Nimbus, I can't stand the default margins, which is why it was originally hardcoded. 

Maybe one day, something a lot more fancy will be possible.

**Comment by dorkbox** [comment]
Fixed via: 0e5d52352b68937ce07fd72c16d0fd2cd4359076

Devs can customize the L&F of the Swing components on the SystemTray separately from the global L&F of Java.

**Comment by tresf** [comment]
@dorkbox you've closed this issue but explicitly tagged myself and @Vzor- for testing over in #22.

> The last remaining issue was #45 before the 3.0 release. Just checking with @Vzor- and @tresf to verify.

Do you have a binary prepared for testing this?  Alternately, I'd be happy to build from source with a a few steps.

**Comment by dorkbox** [comment]
RC10 (on the server) for a "FAT" jar, 3.0 (which has been updated) for the release version.

**Comment by tresf** [comment]
Please reopen.

![image](https://cloud.githubusercontent.com/assets/6345473/23195215/0ca14c5e-f880-11e6-806b-4f0ff42837ac.png)

Something is still changing defaults.

**Comment by dorkbox** [comment]
I suspect is what you are seeing are the HiDPI scaling affect from Win10, and I worked pretty extensively with @Vzor- to figure out appropriate scaling in Win10, since Java elements don't scale at all.

The menu height and left-indent will changed based on the size of the menu image, which is why they are larger in the Dorkbox menu vs "normal" menu. If there is no scaling, on HiDPI settings, the menu doesn't look quite right. See issue #43 

Maybe all 3 of us should chat about what this should look like?

**Comment by dorkbox** [comment]
I'll release 3.1 right now (which fixes some logging + mac menus), then we can address this.

**Comment by tresf** [comment]
> I worked pretty extensively with @Vzor- to figure out appropriate scaling in Win10, since Java elements don't scale at all.

Can you please show evidence of this statement?  Here's what I'm seeing on Windows 10 + Swing.

![image](https://cloud.githubusercontent.com/assets/6345473/23218367/15b31934-f8ea-11e6-84a8-b66039dae729.png)

I can confirm that the icons aren't scaling, but that may be a problem in our code, I would have to investigate to know more.

**Comment by dorkbox** [comment]
Issue #43 documented this this process with @Vzor-

Also DPI issues and workarounds are [here for the workaround](http://stackoverflow.com/questions/30555401/java-disable-dpi-aware-not-working/39372897#39372897) and [here](https://bugs.openjdk.java.net/browse/JDK-8055453) and [here](https://bugs.openjdk.java.net/browse/JDK-8080153) for bug reports. Of note, this is a regression, and the last time DPI scaling worked properly for Swing was in 6u43.

**Comment by tresf** [comment]
> here (https://bugs.openjdk.java.net/browse/JDK-8080153)

That mentions DPI awareness, but is about disabling it, which is irrelevant.

> and here (https://bugs.openjdk.java.net/browse/JDK-8055453)

That's a per-monitor bug.  Qt has some similar problems.  Again, this is an edge case.

> Of note, this is a regression, and the last time DPI scaling worked properly for Swing was in 6u43.

My screenshots show otherwise.

**Comment by dorkbox** [comment]
> My screenshots show otherwise.

And yet the images don't scale? Maybe there is something else we should do, but the problem is that java doesn't scale swing components correctly, and the text size doesn't always line up with what the DPI settings specify.

> That mentions DPI awareness, but is about disabling it, which is irrelevant.

Actually, very relevant, since DPI scaling has to be disabled to fix the issues you are having, where the images are the incorrect size (according to the discussion. I have not personally verified their fix). Currently Java registers as DPI aware, but doesn't scale all swing elements correctly, as your screenshots show. If you disable DPI awareness, it fixes the problem by letting the OS scale everything inside the display manager instead of letting Java do it.

JetBrains also had to release their own JRE bundled with IntelliJ in order to fix this issue. [Here](https://blog.jetbrains.com/idea/2015/07/intellij-idea-15-eap-comes-with-true-hidpi-support-for-windows-and-linux/) is their release notes mentioning scaling for text and icons.

**Comment by tresf** [comment]
> JetBrains also had to release their own JRE bundled with IntelliJ in order to fix this issue. Here is their release notes mentioning scaling for text and icons.

Thanks.  I still don't see a direct correlation.  To quote the article... 

>  The support was somewhat limited and only allowed to use a 2x scale factor (and some displays do have other scale factors, so some of them were not supported.)

My screenshots do not show a 2x scale factor, so that write-up shouldn't be taken verbatim as to the severity of the issues.

Here's a comparison of sorts... 

![image](https://cloud.githubusercontent.com/assets/6345473/23224652/377e5886-f8fd-11e6-8e5a-0c48cc7d5e21.png)


![image](https://cloud.githubusercontent.com/assets/6345473/23224693/55f46d8c-f8fd-11e6-8892-625ff2da3cb1.png)

What I'm seeing is a spacing that is "sane" and a font that is identical by default, running 150%.  I have put no time into researching the icons, but the menu itself looks just fine from where I'm standing.

**Comment by dorkbox** [comment]
I feel like we're going in circles. 

 - @Vzor- reported that text and images are not scaling correctly with DPI changes across different windows machines.

 - The SystemTray now scales according to DPI changes (so 125% DPI increase means the image is 125% larger), and text scaling is based on a "best fit" logic, where a util class measures the height of the rendered font, and chooses a font that is the largest, but still smaller than the requested DPI increase.

 - @Vzor- verified this is now fixed across two different Win10 laptops, both with different screen resolution and default DPI settings.

So now what? Should the images and text not scale based on DPI? I dislike reverting a bug fix that was solved in issue #43. It might be that your machine looks correct by default, but on the two machines that were tested, it absolutely *was not* correct. Perhaps issue #43 should be reopened so we can run more test cases and add exceptions to how scaling works? Maybe it's possible that the default image size is incorrectly set by the utils (ie: it's not 24px tall, but 18px tall)?

**Comment by tresf** [comment]
> @Vzor- reported that text and images are not scaling correctly with DPI changes across different windows machines.

Using Nimbus?  Probably.

> @Vzor- verified this is now fixed across two different Win10 laptops, both with different screen resolution and default DPI settings.

Using Nimbus?  Then his test results were flawed.

> Should the images and text not scale based on DPI?

Text already is.  See test results above.  I'm not sure about images.

>  I dislike reverting a bug fix that was solved in issue #43.

Why though?  Aren't sane defaults better, always?

> Perhaps issue #43 should be reopened

Perhaps, or a new bug should be filed, but only if it's valid. The evidence is in the above posts.  The "A" is pixel-for-pixel exact, by default, using Swing.  If you fail to acknowledge that, I give up. 😸

**Comment by dorkbox** [comment]
**NIMBUS**! Argggh.

Yes, I think you're absolutely right - I wasn't able to see the forest through trees... And that will *absolutely* change the default icon size as well, which I suspect is what is happening. The default image size for the menu entries was set when Nimbus was the L&F... so even now that it uses the defined L&F, the images are now the wrong size.

I will set the default image size based on what the current UI L&F is, instead of an arbitrary value.

Would it be possible for you to do the same font size comparison running v3.1? It might be the font scaling is fine as-is, and just the image is off. It might be both are off. Image and font scaling were the problems we had -- so it'd probably be good to have kyle test on his two machines again once.

**Comment by tresf** [comment]
> Would it be possible for you to do the same font size comparison running v3.1

Sure, do you have a binary to test with or some basic build instructions?

**Comment by tresf** [comment]
Side, note, if you ever want to know your download metrics:  https://api.github.com/repos/dorkbox/SystemTray/releases (search for `download_count`)

**Comment by dorkbox** [comment]
Ah cool.

**Comment by dorkbox** [comment]
Just uploaded 3.1 to releases.

**Comment by dorkbox** [comment]
(still working on build instructions, btw)

**Comment by tresf** [comment]
![image](https://cloud.githubusercontent.com/assets/6345473/23237282/aaef219c-f92a-11e6-8c37-c076c7fd1e9a.png)

(please ignore the Separator differences, that is our code).

**Comment by tresf** [comment]
Here's a better illustration.... It uses the Volume menu as a baseline font comparison... 

Note, the screenshots are from 175% zoom in this test case so please forgive the slight differences between use-cases as well as the natural blurring that occurs when using clipboard at this resolution.

As you can see, Swing has the font exactly.  I'm using the word "Open" in our menu as the test since it is the exact same word that appears in the volume mixer.

![image](https://cloud.githubusercontent.com/assets/6345473/23237639/46cddd0a-f92c-11e6-860c-46e1f7915904.png)

**Comment by dorkbox** [comment]
Wow. I've completely removed all font scaling [here](https://files2.dorkbox.com/SystemTray/system_tray_v3.2_beta.jar). Can you see what this does? I hope it looks correct, then I can finish image scaling.

**Comment by dorkbox** [comment]
note: if it matches, and I hope it does - there's no need to post screenshots.

**Comment by tresf** [comment]
Yes, that fixes it.

**Comment by dorkbox** [comment]
Alright. I hate undoing work, but I cannot argue against what is correct... I really do thing Nimbus is to blame for the issues-that-is-not-really-an-issue...

I'll fix image scaling next.

**Comment by dorkbox** [comment]
fonts are fixed in: ee5f509e7652dae7d489a96407975cb79556a0fc

**Comment by dorkbox** [comment]
Wow, this ended up a lot more complicated than I thought...

The way I'm getting the size of the menu item, is to actually create the `JMenuItem`, then query for it's size and margins. Since the text is correct, I just set the image to be (as close as possible, that looks good) the same size as the text. In my tests, this resulted in correct looking icons, however can you verify? I have a test jar on my server (the same place as before), if you could verify it's now working correctly.

Just so you have an idea, here's the code-snippet that figures out the size of the image for a JMenuItem.

````java
// of note, components can ALSO have different sizes attached to them!
// mini
//      myButton.putClientProperty("JComponent.sizeVariant", "mini");
// small
//      mySlider.putClientProperty("JComponent.sizeVariant", "small");
// large
//      myTextField.putClientProperty("JComponent.sizeVariant", "large");
//
JMenuItem jMenuItem = new JMenuItem("`Tj|┃"); // `Tj|┃ are glyphs that are at the top/bottom of the fontset (usually)

// do the same modifications that would also happen (if specified) for the actual displayed menu items
if (SystemTray.SWING_UI != null) {
    jMenuItem.setUI(SystemTray.SWING_UI.getItemUI(jMenuItem, null));
}
Dimension preferredSize = jMenuItem.getPreferredSize();
Insets margin = jMenuItem.getMargin();

int requestedSize = (int) (preferredSize.getHeight() - margin.bottom - margin.top);
// default values for system L&F:
// ubuntu (metal) : 15px (16px adjusted)
// windows XP (windows) : 15px (16px adjusted) @ 96dpi
// windows 7 (windows) : 15px (16px adjusted)  @ 96dpi, 21px (22px adjusted)  @ 144dpi
// windows 8.1 (windows) : 22px @ 96dpi, 24px @ 120dpi, 31px (30px adjusted) @ 144dpi
// windows 10 (windows) : 22px @ 96dpi, 24px @ 120dpi, 31px (30px adjusted) @ 144dpi

// if within a close range to 16px, it is preferred, since icons will look the absolute best when scaled to 
// 17 -> 16
// all others -> even (ok) or odd (bad)
````

**Comment by tresf** [comment]
The updated jar seems to overcompensate for icon size.


BEFORE:
![image](https://cloud.githubusercontent.com/assets/6345473/23347696/14072e8c-fc72-11e6-8a73-d4149ada5b8f.png)

AFTER:
![image](https://cloud.githubusercontent.com/assets/6345473/23347613/71a1dbe2-fc71-11e6-9f97-70a8242c37f8.png)

BASELINE (Google Chrome Context menu):
![image](https://cloud.githubusercontent.com/assets/6345473/23347734/4208bfc6-fc72-11e6-89d1-98d7baf06dc5.png)

**Comment by tresf** [comment]
Comparing it to the Chrome and Notepad++ icons, they're scaling from `16px` to `20px` whereas your logic scales from `16px` to `24px`.

Interestingly enough:
 * `20px` would technically be 125% zoom.
 * `24px` would technically be 150% zoom.

My screen is set to 175% zoom, so the math -- even with chrome -- works out quite funny.  This tells me there's some additional hardware scaling happening as well which cannot be measured by counting pixels, so the 175% factor must be taken with a grain of salt in our use-cases.  It cannot be taken at face value.

**Comment by dorkbox** [comment]
Hmm, it seems like there is more than just the margins to account for, padding perhaps?  Thanks for the feedback, I'll see if there is more adjustments that should happen.

**Comment by dorkbox** [comment]
So I think I found a way to query the primary monitor **hardware** DPI -- which means I would have a better idea (combined with the logical DPI setting) of what scale to adjust the images to. This won't fix the issue for multi-monitor setups, where different monitors have different scaling, but this will solve the problem (I hope) for most windows users that have scaled displays.

Can you run the (newly uploaded) jar and tell me what the `logical DPI` and `hardware DPI` are in the log? I suspect they might be different on your laptop, since the screenshots that you took have a more pronounced scaling different than on my VMs.

**Comment by tresf** [comment]
> I would have to investigate our icon scaling code to know more about them specifically.

@dorkbox have we done any research to make sure that Java isn't already handling this properly?  Can we just sanely use the natural tray icon size for the menu and remove the complex logic?

> Can you run the (newly uploaded) jar and tell me what the logical DPI and hardware DPI are in the log?

I can yes however I also would like to know your opinion on the above.  It may not be your problem to determine the icon size as all other Swing components would suffer this same problem.

**Comment by dorkbox** [comment]
> @dorkbox have we done any research to make sure that Java isn't already handling this properly? Can we just sanely use the natural tray icon size for the menu and remove the complex logic?

@tresf From what I can tell, java isn't scaling the icon at all. I have another file up (noscale.jar) that removes all scaling from the menu items, so you can see this first-hand.

>  It may not be your problem to determine the icon size as all other Swing components would suffer this same problem.

Naturally, other swing components face this problem, but I honestly don't think Oracle has the will nor the time to fix this until who-knows-when.  This reason is exactly why JetBrains released their own JRE (to fix HiDPI issues) for IntelliJ almost 2 years ago, and why companies like Azul exist (one reason at least it to provide a better GC than Oracle).

I cannot fix other peoples applications; but I can fix the SystemTray deficiencies and make it behave and look correct. I can also provide an API for other developers to query and get the correct scaling for the display, thus enabling them to easily correct the problems they might have through-out their application.

Java 6/7 are [EOL already](http://www.oracle.com/technetwork/java/eol-135779.html), and Java 8 EOL is [September 2017](http://www.oracle.com/technetwork/java/eol-135779.html) for public updates (and I target Java 6+), this means that the public will never see this fixed for Java 6/7 and  and I doubt this will get fixed for Java 8. You might be able to pressure them to fix the scaling issue before Java 8 EOL, but for the vast majority of developers we don't have that kind of agreement with Oracle.

**Comment by tresf** [comment]
> @tresf From what I can tell, java isn't scaling the icon at all

That's not the question.  The question is **should it be** scaling it?  If we can detect and instruct what size to put in the menu we can leave the onus on the developer.

**Comment by tresf** [comment]
My point is, no one wants blurry icons and the 16x16 look just fine.  If we can detect the actual size, the developer should be responsible for pixel-perfect icons or scaling thereof.

**Comment by tresf** [comment]
> You might be able to pressure them to fix the scaling issue before Java 8 EOL,

No, they won't take it if it's not critical, but I think we're missing the point.  We can setup 20x20 icons ourselves if we know that's what needed.  The scaling logic is a noble cause, but spinning our wheels in place without researching this can be wasteful.  If we can determine 20x20, we can scale ourselves.  Pixel perfect applications will need to do this regardless especially considering there's no native SVG support.

What I can do is perform a few unit tests and see if I can fetch 20x20.  I think you're already calculating it for the tray icon (Mac and Linux actually scale that properly without your code) so I think it's a matter of exposing your icon calculation publicly.

**Comment by dorkbox** [comment]
> Any JMenu height/font/DPI/etc guesswork is just redundant if we can already get 20x20 from your library.

True. I'm auto-scaling images, but it's trivial (and preferable) to disable that once the API is in place, since a 22px native image will always look better than a 16px image scaled to 22px.  

Can you see what the `hardware` and `logical` DPI values are for your system? Once we know that is working correctly, then providing an API is trivial.

**Comment by dorkbox** [comment]
Oh, also -- windows is properly scaling the tray icon, and there's no way I know of to get the post-scaled size.

**Comment by tresf** [comment]
(this post was accidentally placed in #46)

@dorkbox so I ran some tests.

Turns out the `16x16` assumption was wrong from the beginning.  This is what I discovered.

100% scale factor:
 * `JMenuItem` is at natural size with an icon `10px` or smaller

125% scale factor:
* `JMenuItem` is at a natural size with an icon of `15px` or smaller

This explains why the `16x16 `icons aren't horrible in our use-cases.  They actually fit better within the menu on HiDPI than they do on a normal DPI.

This can be observed by measuring the distance between Menu items.  In our case, "Automatically start" does not have a 16x16 icon and it appears a few pixels shorter than the rest of the items.

Unit test:
```java
// Echo a JMenuItem height as influenced by the contained ImageIcon to determine the "natural" icon size.
JFrame frame = new JFrame();
frame.setSize(100, 100);
JPopupMenu popup = new JPopupMenu();
popup.setSize(new Dimension(100, 100));
JMenuItem item = new JMenuItem("foo bar");
popup.add(item);

ArrayList<ImageIcon> icons = new ArrayList<>();
for (int i = 1; i < 128; i++) {
    // create a 1px x n-px image to set as the icon
    icons.add(new ImageIcon(new BufferedImage(1, i, BufferedImage.TYPE_BYTE_BINARY)));
}
frame.setVisible(true);
popup.show(frame, 50, 50);

for (ImageIcon i : icons) {
    item.setIcon(i);
    item.revalidate();
    // revalidate should be enough, but it's not
    popup.setVisible(false);
    popup.show(frame, 50, 50);
    System.out.println(i.getIconHeight() + "x" + i.getIconHeight() + " icon \t>>>>>>>>> " + item.getHeight() + "px tall item");
}

```


With 100% zoom, I get this:

```diff
+7x7 icon 	>>>>>>>>> 16px tall item
+8x8 icon 	>>>>>>>>> 16px tall item
+9x9 icon 	>>>>>>>>> 16px tall item
+10x10 icon 	>>>>>>>>> 16px tall item
-11x11 icon 	>>>>>>>>> 17px tall item
-12x12 icon 	>>>>>>>>> 18px tall item
```

With 125% zoom, I get this:

```diff
+14x14 icon 	>>>>>>>>> 21px tall item
+15x15 icon 	>>>>>>>>> 21px tall item
-16x16 icon 	>>>>>>>>> 22px tall item
-17x17 icon 	>>>>>>>>> 23px tall item
-18x18 icon 	>>>>>>>>> 24px tall item
```

How does this help us?  Well, it further proves that we should be careful what assumptions we make when scaling things for end-users.

I still like the idea of the menu icons being the same size as the tray icon and I'll test the scaling logic next, but wanted to share these findings.

**Comment by tresf** [comment]
In regards to testing the latest build ([`system_tray_v3.2_beta.jar`](https://files2.dorkbox.com/SystemTray/system_tray_v3.2_beta.jar)), the icon sizes don't seem to have changed from previous release.

My baseline is Google Chrome.

 * Chrome scales 16x16 to 20x20 (Showtime logo)
 * Dorkbox scales 16x16 to 24x24 (List item logo)


> ![image](https://cloud.githubusercontent.com/assets/6345473/23393566/c2663aec-fd51-11e6-8fd8-6f82b900e3dd.png)

**Comment by dorkbox** [comment]
I need the `hardware` and `logical` DPI values from the log.

I suspect there is hardware scaling on your system, since non-scaled systems return different results than what you posted.

**Comment by tresf** [comment]
I looked for some logs but couldn't find anything excellent a font size in  the logs.  I'd happily try again .

**Comment by dorkbox** [comment]
It should be a few lines up. It might be an issue with logging -- so I just uploaded a standalone `system_tray_v3.1_normal.jar` to the server. Can you run that (`java -jar system_tray_v3.1_normal.jar`)? The info should show up in the log.

**Comment by tresf** [comment]
@dorkbox this is what I get:

```log
Error: Invalid or corrupt jarfile C:\Users\lcladmin\Downloads\system_tray_v3.2_beta.3.jar
```

**Comment by dorkbox** [comment]
Have to run `system_tray_v3.1_normal.jar` instead. `system_tray_v3.2_beta.3.jar` is not an executable jar (it's just the library w/ no deps)

**Comment by tresf** [comment]
> system_tray_v3.2_beta.jar
>>  system_tray_v3.1_normal.jar

😕 Ok... :)

Disclaimer, this is on a Windows 10 VM running 125% zoom.  I won't be at the true HiDPI until much later.

```log
C:\%userprofile%>java -jar %userprofile%\Downloads\system_tray_v3.1_normal.jar
10:05:16.273 [main] DEBUG dorkbox.systemTray.SystemTray - OS: Windows 10
10:05:16.375 [main] DEBUG dorkbox.systemTray.SystemTray - Arch: x86
10:05:16.380 [main] DEBUG dorkbox.systemTray.SystemTray - Oracle Corporation Java HotSpot(TM) Client VM 1.8.0_65
10:05:16.394 [main] DEBUG dorkbox.systemTray.SystemTray - Is AutoTraySize? true
10:05:16.425 [main] DEBUG dorkbox.systemTray.SystemTray - Is JavaFX detected? false
10:05:16.431 [main] DEBUG dorkbox.systemTray.SystemTray - Is SWT detected? false
10:05:16.436 [main] DEBUG dorkbox.systemTray.SystemTray - Forced tray type: Swing
10:05:16.452 [main] DEBUG dorkbox.systemTray.SystemTray - FORCE_GTK2: false
10:05:17.555 [main] DEBUG dorkbox.systemTray.SystemTray - Successfully changed tray icon size to: 0
10:05:17.596 [main] DEBUG dorkbox.systemTray.SystemTray - Windows version: '[10, 0]'
10:05:17.649 [main] DEBUG dorkbox.systemTray.SystemTray - Windows hardware DPI: '120'
10:05:17.682 [main] DEBUG dorkbox.systemTray.SystemTray - Windows logical DPI: '120'
FONT HEIGHT 19
10:05:18.188 [main] DEBUG dorkbox.systemTray.SystemTray - Swing L&F: Windows
10:05:18.194 [main] DEBUG dorkbox.systemTray.SystemTray - Swing L&F menu item height w/o margins: 24 adjusted to 24
10:05:18.199 [main] DEBUG dorkbox.systemTray.SystemTray - ScalingFactor is '4.0', tray icon size is '64', menu item size is '16'.
10:05:18.514 [main] INFO dorkbox.systemTray.SystemTray - Successfully loaded type: _SwingTray
```

**Comment by dorkbox** [comment]
What happens if you run at 175% (or 100%) zoom?

I'm curious to see if the `Windows hardware DPI: '120'` ever changes. On my VM's it always matches the logical DPI, and I'm hoping on your system it's different.

**Comment by tresf** [comment]
@dorkbox a collegue was nice enough to run this on his 125% HiDPI laptop for me.

Looks to be identical.

```log
11:13:20.318 [main] DEBUG dorkbox.systemTray.SystemTray - OS: Windows 10
11:13:20.346 [main] DEBUG dorkbox.systemTray.SystemTray - Arch: x86
11:13:20.347 [main] DEBUG dorkbox.systemTray.SystemTray - Oracle Corporation Java HotSpot(TM) Client VM 1.8.0_121
11:13:20.347 [main] DEBUG dorkbox.systemTray.SystemTray - Is AutoTraySize? true
11:13:20.347 [main] DEBUG dorkbox.systemTray.SystemTray - Is JavaFX detected? false
11:13:20.347 [main] DEBUG dorkbox.systemTray.SystemTray - Is SWT detected? false
11:13:20.348 [main] DEBUG dorkbox.systemTray.SystemTray - Forced tray type: Swing
11:13:20.348 [main] DEBUG dorkbox.systemTray.SystemTray - FORCE_GTK2: false
11:13:20.935 [main] DEBUG dorkbox.systemTray.SystemTray - Successfully changed tray icon size to: 0
11:13:20.947 [main] DEBUG dorkbox.systemTray.SystemTray - Windows version: '[10, 0]'
11:13:20.967 [main] DEBUG dorkbox.systemTray.SystemTray - Windows hardware DPI: '120'
11:13:20.968 [main] DEBUG dorkbox.systemTray.SystemTray - Windows logical DPI: '120'
FONT HEIGHT 19
11:13:21.121 [main] DEBUG dorkbox.systemTray.SystemTray - Swing L&F: Windows
11:13:21.121 [main] DEBUG dorkbox.systemTray.SystemTray - Swing L&F menu item height w/o margins: 24 adjusted to 24
11:13:21.122 [main] DEBUG dorkbox.systemTray.SystemTray - ScalingFactor is '4.0', tray icon size is '64', menu item size is '16'.
11:13:21.242 [main] INFO dorkbox.systemTray.SystemTray - Successfully loaded type: _SwingTray
```

**Comment by dorkbox** [comment]
What about at 175% and 100% zoom?

**Comment by tresf** [comment]
Running 100% on a Surface Pro 3, but with internal display disabled; standard monitors:

```log
13:24:41.032 [main] DEBUG dorkbox.systemTray.SystemTray - OS: Windows 10
13:24:41.036 [main] DEBUG dorkbox.systemTray.SystemTray - Arch: amd64
13:24:41.037 [main] DEBUG dorkbox.systemTray.SystemTray - Oracle Corporation Java HotSpot(TM) 64-Bit Server VM 1.8.0_121
13:24:41.037 [main] DEBUG dorkbox.systemTray.SystemTray - Is AutoTraySize? true
13:24:41.037 [main] DEBUG dorkbox.systemTray.SystemTray - Is JavaFX detected? false
13:24:41.037 [main] DEBUG dorkbox.systemTray.SystemTray - Is SWT detected? false
13:24:41.037 [main] DEBUG dorkbox.systemTray.SystemTray - Forced tray type: Swing
13:24:41.037 [main] DEBUG dorkbox.systemTray.SystemTray - FORCE_GTK2: false
13:24:41.236 [main] DEBUG dorkbox.systemTray.SystemTray - Successfully changed tray icon size to: 0
13:24:41.237 [main] DEBUG dorkbox.systemTray.SystemTray - Windows version: '[10, 0]'
13:24:41.246 [main] DEBUG dorkbox.systemTray.SystemTray - Windows hardware DPI: '96'
13:24:41.246 [main] DEBUG dorkbox.systemTray.SystemTray - Windows logical DPI: '96'
FONT HEIGHT 15
13:24:41.392 [main] DEBUG dorkbox.systemTray.SystemTray - Swing L&F: Windows
13:24:41.393 [main] DEBUG dorkbox.systemTray.SystemTray - Swing L&F menu item height w/o margins: 22 adjusted to 22
13:24:41.393 [main] DEBUG dorkbox.systemTray.SystemTray - ScalingFactor is '4.0', tray icon size is '64', menu item size is '16'.
13:24:41.495 [main] INFO dorkbox.systemTray.SystemTray - Successfully loaded type: _SwingTray
```

**Comment by dorkbox** [comment]
What about at 175% and 100% zoom?

**Comment by tresf** [comment]
@dorkbox the physical monitor isn't showing a difference over the VM, so it should ultimately match your VM testing.  If that's insufficient for test results, I will have access to a HiDPI machine later and can provide those results as well.

**Comment by tresf** [comment]
175% on XPS 13

```log
21:12:23.242 [main] DEBUG dorkbox.systemTray.SystemTray - OS: Windows 10
21:12:23.248 [main] DEBUG dorkbox.systemTray.SystemTray - Arch: amd64
21:12:23.249 [main] DEBUG dorkbox.systemTray.SystemTray - Oracle Corporation Java HotSpot(TM)

21:12:23.249 [main] DEBUG dorkbox.systemTray.SystemTray - Is AutoTraySize? true
21:12:23.250 [main] DEBUG dorkbox.systemTray.SystemTray - Is JavaFX detected? false
21:12:23.250 [main] DEBUG dorkbox.systemTray.SystemTray - Is SWT detected? false
21:12:23.250 [main] DEBUG dorkbox.systemTray.SystemTray - Forced tray type: Swing
21:12:23.250 [main] DEBUG dorkbox.systemTray.SystemTray - FORCE_GTK2: false
21:12:23.474 [main] DEBUG dorkbox.systemTray.SystemTray - Successfully changed tray icon size
21:12:23.476 [main] DEBUG dorkbox.systemTray.SystemTray - Windows version: '[10, 0]'
21:12:23.485 [main] DEBUG dorkbox.systemTray.SystemTray - Windows hardware DPI: '168'
21:12:23.486 [main] DEBUG dorkbox.systemTray.SystemTray - Windows logical DPI: '168'
FONT HEIGHT 27
21:12:23.556 [main] DEBUG dorkbox.systemTray.SystemTray - Swing L&F: Windows
21:12:23.556 [main] DEBUG dorkbox.systemTray.SystemTray - Swing L&F menu item height w/o margi
21:12:23.557 [main] DEBUG dorkbox.systemTray.SystemTray - ScalingFactor is '4.0', tray icon size is '16'.
21:12:23.623 [main] INFO dorkbox.systemTray.SystemTray - Successfully loaded type: _SwingTray
```

**Comment by tresf** [comment]
175%, Dell XPS 13 Dev Edition

```log
21:17:00.409 [main] DEBUG dorkbox.systemTray.SystemTray - OS: Windows 10
21:17:00.436 [main] DEBUG dorkbox.systemTray.SystemTray - Arch: amd64
21:17:00.467 [main] DEBUG dorkbox.systemTray.SystemTray - Oracle Corporation Java HotSpot(TM) 64-Bit Server VM 1.8.0_111
21:17:00.494 [main] DEBUG dorkbox.systemTray.SystemTray - Is AutoTraySize? true
21:17:00.508 [main] DEBUG dorkbox.systemTray.SystemTray - Is JavaFX detected? false
21:17:00.521 [main] DEBUG dorkbox.systemTray.SystemTray - Is SWT detected? false
21:17:00.523 [main] DEBUG dorkbox.systemTray.SystemTray - Forced tray type: Swing
21:17:00.527 [main] DEBUG dorkbox.systemTray.SystemTray - FORCE_GTK2: false
21:17:01.003 [main] DEBUG dorkbox.systemTray.SystemTray - Successfully changed tray icon size to: 0
21:17:01.005 [main] DEBUG dorkbox.systemTray.SystemTray - Windows version: '[10, 0]'
21:17:01.020 [main] DEBUG dorkbox.systemTray.SystemTray - Windows hardware DPI: '168'
21:17:01.020 [main] DEBUG dorkbox.systemTray.SystemTray - Windows logical DPI: '168'
FONT HEIGHT 27
21:17:01.170 [main] DEBUG dorkbox.systemTray.SystemTray - Swing L&F: Windows
21:17:01.175 [main] DEBUG dorkbox.systemTray.SystemTray - Swing L&F menu item height w/o margins: 31 adjusted to 30
21:17:01.175 [main] DEBUG dorkbox.systemTray.SystemTray - ScalingFactor is '4.0', tray icon size is '64', menu item size is '16'.
21:17:01.281 [main] INFO dorkbox.systemTray.SystemTray - Successfully loaded type: _SwingTray
```

**Comment by tresf** [comment]
Even setting desktop zoom to one value and custom font scaling to another, the Hardware and Logical DPI are the same.

I can probably fool this value by running a secondary monitor, but I'm not sure which monitor Java will attach to.

**Comment by dorkbox** [comment]
@tresf I have an updated version on my server that uses a different value to check for the hardware DPI -- can you see if it's anything other than `96` (on my VM that is zoomed, hardware is `96` and logical is `144`).  

If it is different for you (it should stay the same value per machines across all zoom levels), then it means I'm now passing the correct value to the windows API and this issue is almost resolved.

**Comment by tresf** [comment]
125% on physical hardware with updated JAR that uses a different value to check for the hardware DPI:

```log
10:22:33.993 [main] DEBUG dorkbox.systemTray.SystemTray - OS: Windows 10
10:22:34.035 [main] DEBUG dorkbox.systemTray.SystemTray - Arch: x86
10:22:34.036 [main] DEBUG dorkbox.systemTray.SystemTray - Oracle Corporation Java HotSpot(TM) Client VM 1.8.0_121
10:22:34.036 [main] DEBUG dorkbox.systemTray.SystemTray - Is AutoTraySize? true
10:22:34.037 [main] DEBUG dorkbox.systemTray.SystemTray - Is JavaFX detected? false
10:22:34.037 [main] DEBUG dorkbox.systemTray.SystemTray - Is SWT detected? false
10:22:34.037 [main] DEBUG dorkbox.systemTray.SystemTray - Forced tray type: Swing
10:22:34.037 [main] DEBUG dorkbox.systemTray.SystemTray - FORCE_GTK2: false
10:22:34.991 [main] DEBUG dorkbox.systemTray.SystemTray - Successfully changed tray icon size to: 0
10:22:34.997 [main] DEBUG dorkbox.systemTray.SystemTray - Windows version: '[10, 0]'
10:22:35.026 [main] DEBUG dorkbox.systemTray.SystemTray - Windows hardware DPI: '125'
10:22:35.027 [main] DEBUG dorkbox.systemTray.SystemTray - Windows logical DPI: '120'
FONT HEIGHT 19
10:22:35.335 [main] DEBUG dorkbox.systemTray.SystemTray - Swing L&F: Windows
10:22:35.336 [main] DEBUG dorkbox.systemTray.SystemTray - Swing L&F menu item height w/o margins: 24 adjusted to 24
10:22:35.337 [main] DEBUG dorkbox.systemTray.SystemTray - ScalingFactor is '4.0', tray icon size is '64', menu item size is '16'.
10:22:35.560 [main] INFO dorkbox.systemTray.SystemTray - Successfully loaded type: _SwingTray
```

**Comment by dorkbox** [comment]
Out of curiosity, what are the values at different zoom levels on that machine?

**Comment by tresf** [comment]
125% on Dell XPS 13 Dev edition... 

```
22:36:43.143 [main] DEBUG dorkbox.systemTray.SystemTray - OS: Windows 10
22:36:43.159 [main] DEBUG dorkbox.systemTray.SystemTray - Arch: amd64
22:36:43.162 [main] DEBUG dorkbox.systemTray.SystemTray - Oracle Corporation Java HotSpot(TM) 64-Bit Server VM 1.8.0_111

22:36:43.163 [main] DEBUG dorkbox.systemTray.SystemTray - Is AutoTraySize? true
22:36:43.163 [main] DEBUG dorkbox.systemTray.SystemTray - Is JavaFX detected? false
22:36:43.163 [main] DEBUG dorkbox.systemTray.SystemTray - Is SWT detected? false
22:36:43.164 [main] DEBUG dorkbox.systemTray.SystemTray - Forced tray type: Swing
22:36:43.164 [main] DEBUG dorkbox.systemTray.SystemTray - FORCE_GTK2: false
22:36:43.652 [main] DEBUG dorkbox.systemTray.SystemTray - Successfully changed tray icon size to: 0
22:36:43.656 [main] DEBUG dorkbox.systemTray.SystemTray - Windows version: '[10, 0]'
22:36:43.676 [main] DEBUG dorkbox.systemTray.SystemTray - Windows hardware DPI: '145'
22:36:43.676 [main] DEBUG dorkbox.systemTray.SystemTray - Windows logical DPI: '120'
FONT HEIGHT 19
22:36:43.850 [main] DEBUG dorkbox.systemTray.SystemTray - Swing L&F: Windows
22:36:43.851 [main] DEBUG dorkbox.systemTray.SystemTray - Swing L&F menu item height w/o margins: 24 adjusted to 24
22:36:43.852 [main] DEBUG dorkbox.systemTray.SystemTray - ScalingFactor is '4.0', tray icon size is '64', menu item size
 is '16'.
22:36:44.018 [main] INFO dorkbox.systemTray.SystemTray - Successfully loaded type: _SwingTray
```

**Comment by tresf** [comment]
175% on Dell XPS 13 Dev edition...

```
22:41:24.392 [main] DEBUG dorkbox.systemTray.SystemTray - OS: Windows 10
22:41:24.400 [main] DEBUG dorkbox.systemTray.SystemTray - Arch: amd64
22:41:24.401 [main] DEBUG dorkbox.systemTray.SystemTray - Oracle Corporation Java HotSpot(TM) 64-Bit Server VM 1.8.0_111
22:41:24.401 [main] DEBUG dorkbox.systemTray.SystemTray - Is AutoTraySize? true
22:41:24.401 [main] DEBUG dorkbox.systemTray.SystemTray - Is JavaFX detected? false
22:41:24.402 [main] DEBUG dorkbox.systemTray.SystemTray - Is SWT detected? false
22:41:24.402 [main] DEBUG dorkbox.systemTray.SystemTray - Forced tray type: Swing
22:41:24.402 [main] DEBUG dorkbox.systemTray.SystemTray - FORCE_GTK2: false
22:41:24.651 [main] DEBUG dorkbox.systemTray.SystemTray - Successfully changed tray icon size to: 0
22:41:24.653 [main] DEBUG dorkbox.systemTray.SystemTray - Windows version: '[10, 0]'
22:41:24.662 [main] DEBUG dorkbox.systemTray.SystemTray - Windows hardware DPI: '145'
22:41:24.662 [main] DEBUG dorkbox.systemTray.SystemTray - Windows logical DPI: '168'
FONT HEIGHT 27
22:41:24.740 [main] DEBUG dorkbox.systemTray.SystemTray - Swing L&F: Windows
22:41:24.741 [main] DEBUG dorkbox.systemTray.SystemTray - Swing L&F menu item height w/o margins: 31 adjusted to 30
22:41:24.741 [main] DEBUG dorkbox.systemTray.SystemTray - ScalingFactor is '4.0', tray icon size is '64', menu item size is '16'.
22:41:24.820 [main] INFO dorkbox.systemTray.SystemTray - Successfully loaded type: _SwingTray
```

**Comment by tresf** [comment]
**125%** - 16x16 measures as 20x20
**175%** - 16x16 measures as 28x28

The general equation at a glance seems to be `logical_dpi / 96`.

```java
// 125%
int w = 16;
int h = 16;
int LOGICAL_DPI = 120;

int w2 = w * (logical_dpi / 96);
int h2 = h * (logical_dpi / 96);

// >> w2: 20
// >> h2: 20
```

```
// 175%
int w = 16;
int h = 16;
int LOGICAL_DPI = 168;

int w2 = w * (logical_dpi / 96);
int h2 = h * (logical_dpi / 96);

// >> w2: 28
// >> h2: 28
```

**Comment by dorkbox** [comment]
@tresf many thanks for running the tests.

> The hardware DPI seems to be the physical hardware density for the OS

Yup, `hardware DPI` is supposed to be the physical, hardware DPI reported by the monitor; I just wanted to confirm what it shows on your system. Getting the API call correct was a little tricky.

> I'd guess hardware DPI is useless unless we're writing an operating system.

Not exactly ... the hardware DPI is actually useful, as it tells me if it's a HiDPI screen or not, and the logical DPI tells me what "zoom" level is set in the display settings in windows.

I'm not exactly sure how to use these numbers yet, but my first attempt is going to replace `96` with the hardware DPI, and this way the correct scale can be calculated. I plan on basing this from the tests you ran incrementing the size of the image 1px at a time and measuring the size of the JMenuItem.

My guess is that the closer those two numbers, the more "1:1" what you see on the screen is (in size) compared to my machine which does not have a HiDPI display. This would also explain why a 100% zoom on a 125% display makes everything look tiny, while on a different display it would look 'normal'.


On the Dell XPS 13 Dev edition, what are the icon sizes for Chrome for different zoom levels? This way I can know if I'm getting the scaling algo correct without bugging you every time I have something.

**Comment by tresf** [comment]
> it tells me if it's a HiDPI screen or not

The DPI of the screen is irrelevant.  The only relevance is how the pixels are being scaled.  For example, accessibility mode for the sight impaired could be using 175% zoom on a standard DPI.  Please don't make any assumptions around HiDPI if we can hit the lowest common denominator with logical Zoom.

> I'm not exactly sure how to use these numbers yet

The `LOGICAL_ZOOM / 96` works out, what are your concerns?

> My guess is that the closer those two numbers, the more "1:1" what you see on the screen is (in size) compared to my machine which does not have a HiDPI display.

I disagree.  They should be inversely proportional.

> This would also explain why a 100% zoom on a 125% display makes everything look tiny, while on a different display it would look 'normal'.

There's no such thing as a 125% display.  Rather, the OS determines on install that 125% is a "sane default" based on pixel density.

The reason things look small on an HiDPI display is because the monitor's physical pixel density is higher, so you can fit more pixels on the screen.  HiDPI is just smaller pixels.  It's still the same technology we've used for ever.  The major difference is the OS takes advantage of it and allows the text and icons to be readable.  HiDPI is more of an OS scaling problem.

For example, if you install Windows XP on a Dell XPS 13 Dev edition, it won't auto-increase the font sizes or the icons.  It will just run at regular size (which is a bit too small to use on a daily basis).  Windows 7 did an OK job and Windows 8 and Windows 10 did a better job of scaling the desktop items based on what the screen "Looked like".  We just care what the OS is doing.  We really shouldn't care about what the hardware is capable of.

**Comment by tresf** [comment]
>  I plan on basing this from the tests you ran incrementing the size of the image 1px at a time and measuring the size of the JMenuItem.

Ok, but remember, these tests [failed](https://github.com/dorkbox/SystemTray/issues/45#issuecomment-282949753) with a `16x16` pixel assumption.

**Comment by dorkbox** [comment]
> The DPI of the screen is irrelevant. .... we can hit the lowest common denominator with logical Zoom.

I was using logical zoom exclusively when you noticed the problem with scaling.

> The LOGICAL_ZOOM / 96 works out, what are your concerns?

Something is off with logical zoom, and it does not work correctly. The current code at the time when you first noticed the problem is [here](https://github.com/dorkbox/SystemTray/blob/master/src/dorkbox/systemTray/util/ImageUtils.java#L140). You will notice it is using logical zoom values only.


> We just care what the OS is doing. We really shouldn't care about what the hardware is capable of.

I think we should back up just a tad. We are both experienced engineers, with years of experience.

## What is the problem?
Java is not scaling the images correctly on certain hardware configurations.

### What do we know does not work?
What is present in 3.1 does not work. This code uses logical scaling values to determine the correct image size. Maybe using logical values is correct and something else is wrong? Possibly. Since Java isn't correctly scaling the images, I'm having to dig into C API calls to try to figure it out.


### What is a possible solution?
Maybe the answer is to use just logical values, maybe the answer is to use both logical **and** hardware values. I *do know* that just using the logical values doesn't work, because if it did, we would not be here trying to figure it out. Again, something else might be wrong, and this is why we are going back-and-forth.

Maybe I did something wrong in calculating the logical DPI and scaling values, I'm not sure - but at least we are taking steps forward to figure out what is going on. I know for a fact it works fine on my machine (and VMs) with a hardware DPI of 96, but that's not what we both want -- we want it to work and look correct on machines with different hardware configurations.

**Comment by tresf** [comment]
I simply can't dedicate more time on this issue.  We use `JXTrayIcon` on XP - through 10 without a single complaint.  The major advantage to using dorkbox on Windows was the non-distored tray icon, but I believe a recent Java update has fixed that.  The major advantage on Mac is the right-click menu.

At a glance it appears you're trying to use tray-icon-size assumptions on Swing components which seems flawed logic.  The swing icons can be as large or small as one wishes. 

At this point, I think I'd prefer to turn off scaling altogether and roll the logic into our own image cache.  We bundle pixel-perfect icons anyway so we really just need to ship with `16`, `20`,. `24` and `28` pixel icons and use the LOGICAL_ZOOM logic above and then we can extend this to our Swing components as well.

**Comment by dorkbox** [comment]
Sounds reasonable -- I think you're right, the most straight-forward solution is to disable scaling.

I'll have a release soon that has scaling disabled and provides an API to get hardware (if one needs it) and logical DPI values from windows, and display scaling values for MacOS and Linux (which are not currently exposed).

**Comment by tresf** [comment]
@dorkbox just a friendly bump to see how soon remaining fixes will be available.

**Comment by dorkbox** [comment]
@tresf Should be in about 2 weeks -- I've been traveling a lot this past month, and been unable to get much work done.

**Comment by dorkbox** [comment]
Quick update : still in progress, but now I've passed internal testing and wrapping up some of the pieces.

**Comment by dorkbox** [comment]
Update: Fixing a bug w.r.t Ubuntu 16.04 + using the System L&F

**Comment by dorkbox** [comment]
So good news and bad news. 

Bad news: There is no way to reliably get the information need to make swing (under windows) be the same L&F as native menus.

Good news: I've added a native windows implementation - done getting the icon to show up, WIP on the menu entries. 

Also, I've fixed the Unix/Linux Swing implementation (it's actually uses X11, so it's now the fallback for when AppIndicator/GtkStatusIcon fails). In particular, this is very useful on Ubuntu, where GTK2 apps cannot use AppIndicators without an extra library (and GtkStatusIcon support was removed).

**Comment by tresf** [comment]
> Bad news: There is no way to reliably get the information need to make swing (under windows) be the same L&F as native menus.

Do you mean the Windows 10 ModernUI stuff, or do you mean the Windows font and spacing stuff as defended here https://github.com/dorkbox/SystemTray/issues/45#issuecomment-281746818?

**Comment by dorkbox** [comment]
As for the Windows font and spacing stuff in https://github.com/dorkbox/SystemTray/issues/45#issuecomment-281746818.

**Comment by dorkbox** [comment]
It is possible to get it close, but it's not exact, and the HiDPI stuff *really* buggers up the spacing/size.

**Comment by tresf** [comment]
I'm still confused.  It's dead-on by default.  What is there to code?  Is it the icons that muck it all up?  I don't understand.

**Comment by dorkbox** [comment]
The icons muck it all up.

**Comment by dorkbox** [comment]
The ````SystemTray.AUTO_SIZE=false;```` parameter will keep all artwork at it's provided size. There are 2 more API's added, ````SystemTray.getTrayImageSize()```` and ````SystemTray.getMenuImageSize()````.

For getting the tray/menu image sizes...
- MacOS is done (Apple hardcoded the values)
- Linux menu sizes are done, WIP for the tray size (themes and DE's can change it).
- Windows Tray size is almost done (it's also hardcoded, just regression testing all versions), and menu sizes are all over the place.

For swing, the logic you figured out (from https://github.com/dorkbox/SystemTray/issues/45#issuecomment-282788092) works most of the time, but not for Windows 8.1+ at different DPIs, and to get the actual icon size (as far as I can tell) for the context menu requires jumping into native code. There is already a BSD licensed JNA windows tray implementation, so it's going very, very quickly to add that functionality. 

I've decided to do this because now that the size is no longer hardcoded (it is now getting the largest image size possible before changing the menu size), means that at different DPIs there are pretty different menu heights, which *really* sticks out from the other menus. 

There have been several other issues fixed (one was a race condition for Swing menus), and the other is the checkmark menu entries getting offset in Ubuntu (https://github.com/dorkbox/SystemTray/issues/42). I'll mark that issue fixed once the code has been pushed and it's verified.

**Comment by tresf** [comment]
> works most of the time, but not for Windows 8.1+ at different DPIs, and to get the actual icon size (as far as I can tell)

So you're coding for an edge-case (HiDPI) on and edge-case (Windows 8.1) to avoid what Java has already done?  Isn't this overkill?

**Comment by dorkbox** [comment]
> So you're coding for an edge-case (HiDPI) on and edge-case (Windows 8.1) to avoid what Java has already done? Isn't this overkill?

I don't want to any edge-case stuff, because that would be close to impossible (and overkill) because different themes and settings can change the menu size.

**Comment by dorkbox** [comment]
No edge cases necessary, I found the correct API to get the tray/menu correct image size.

**Comment by dorkbox** [comment]
This (and some-what related issues) are fixed here:
Windows XP: 2eea54bea3a26597d179e7859e659ea2886836e4 (for [this issue](http://bugs.java.com/bugdatabase/view_bug.do?bug_id=4199382))
Windows (General): 2ca5cdf622019f3cf11e89f53e4ee266f390ee48

The system, if `AUTO_SIZE=true` (the default), will automatically resize all icons to what is reported via native methods, the idea being that a single 64x64 image can be provided, then be downsized as applicable. Regardless of `AUTO_SIZE`, one can always get the correct tray/menu image size via `SystemTray.getTrayImageSize()` and `SystemTray.getMenuImageSize()`, after `SystemTray.get()` has been called, but before any images have been assigned to the tray/menu. Of important note, not all versions of Windows have the same image size (it depends on the current theme applied), as is the same for Linux. MacOSX will always have the same image size (based on DPI settings).

Additionally, HiDPI settings are also accounted for, so whatever the reported image size, is the correct image size for the DPI applied. There are "per-monitor" settings for Windows 8.1+, and it is not particularly easy to determine multiple screens and which screen the system tray is on. The overall DPI will be used based on windows handler in which the Java app is launched. I've found that this usually is correct since I'm querying windows directly for what image size to use.

I will post post test/beta jars for validation before the 3.2 release, as image sizes greatly rely on the underlying hardware/software configuration, and should should be fully resolved before release.

**Comment by dorkbox** [comment]
@tresf I've [pushed the beta](https://github.com/dorkbox/SystemTray/releases/tag/3.2-beta) "all in one jar" for testing, if you'd like to verify this issue is resolved for you. Once I hear from you, I'll close this issue and release 3.2

**Comment by tresf** [comment]
@dorkbox we'd be happy to test, but we can't go live without tooltips, so we'd have to split our code for Windows and Mac.

**Comment by dorkbox** [comment]
Hm. Alright, I'll add tooltips. :)

**Comment by dorkbox** [comment]
Release 3.2 is out, with added tooltips.
